content/blog/using-the-gdb4hpc-debugger-at-scale-with-a-cuda-mpi-hpc-application.md

```diff
@@ -0,0 +1,545 @@
+---
+title: Using the gdb4hpc debugger at scale with a CUDA/MPI HPC application
+date: 2025-01-17T16:37:53.003Z
+author: Josh Chandler
+authorimage: /img/e01ld9fh0jz-u014j40c3d2-43d529500caf-192.jpg
+thumbnailimage: ""
+disable: false
+tags:
+  - HPE Cray Programming Environment
+  - gdb4hpc
+  - debuggers
+  - developer tools
+  - HPC
+  - CPE
+  - cray
+---
+Command line debuggers for Linux have existed for decades. The gdb debugger is the most famous and arguably the most powerful. But gdb has a weakness when it comes to HPC applications - it can only run on one system at a time. gdb can't be used to debug an HPC application at scale because HPC applications run tens of thousands of processes on thousands of systems at once!
+
+This is where gdb4hpc comes in. gdb4hpc is part of the HPE Cray Programming Environment package. It captures the powerful features of gdb and allows the user to apply them to an HPC application at scale. gdb4hpc can debug HPC applications running on thousands of nodes.
+
+gdb4hpc works by connecting the user to many instances of gdb at once. gdb4hpc controls all of the instances of gdb at once and aggregates and filters the results into representations that will comfortably fit on a single terminal screen.
+
+<center><img src="/img/gdb4hpc-controlling-gdbs.png" width="95%" alt="Illustration showing how gdb4hpc connects to individual gdb instances. A single gdb4hpc instance is run on the login node. On each compute node, multiple application ranks are running. Each application rank has an instance of gdb attached to it. In turn, gdb4hpc remotely attaches to each individual gdb instance." title="gdb4hpc controlling multiple instances of gdb"></center>
+
+In this tutorial, you will learn how to debug a multinode MPI/CUDA application
+with gdb4hpc in the HPE Cray Programming Environment. This tutorial uses a CUDA
+application and NVIDIA GPUs as examples, but the concepts are applicable to HIP
+applications on AMD GPUs as well.
+
+## Prerequisites
+
+This tutorial assumes you already have a familiarity with command line debuggers, git, MPI, and CUDA.
+
+You will need a system with access to Cray C++ compilers and gdb4hpc, which are parts of the HPC Cray Programming Environment. You will also need access to the NVIDIA C++ compiler.
+
+## Setup
+
+To set up the debugging session, you will need to obtain the sample application
+and compile it for your chosen GPU with debug information.
+
+### Download The Sample Code
+
+For this tutorial, you will use the `simpleMPI` example from
+[NVIDIA's cuda-samples repository](https://github.com/NVIDIA/cuda-samples).
+`simpleMPI` does a simple distributed calculation involving square roots and averages:
```

Comment:
I should say the same thing for "simpleMPI", but since it's in a code block, visually it kind of works this way.